libcontainerd/container_linux.go

```diff
@@ -194,3 +195,18 @@ func (ctr *container) handleEvent(e *containerd.Event) error {
 	}
 	return nil
 }
+
+// discardFifos attempts to fully read the container fifos to unblock processes
+// that may be blocked on the writer side.
+func (ctr *container) discardFifos() {
+	for _, i := range []int{syscall.Stdout, syscall.Stderr} {
+		f := ctr.fifo(i)
+		c := make(chan struct{})
+		go func() {
+			close(c) // this channel is used to not close the writer too early, before readonly open has been called.
```

Comment:
`closeReaderFifo` may still run before `openReaderFromFifo` (even though it is less likely with this synchronization point). Maybe this is better:

		r := openReaderFromFifo(f)
		go io.Copy(ioutil.Discard, r)
		closeReaderFifo(f)